addons/website_sale/static/src/js/website_sale_recently_viewed.js

```diff
@@ -0,0 +1,186 @@
+odoo.define('website_sale.recently_viewed', function (require) {
+
+var concurrency = require('web.concurrency');
+var config = require('web.config');
+var core = require('web.core');
+var publicWidget = require('web.public.widget');
+var utils = require('web.utils');
+var wSaleUtils = require('website_sale.utils');
+
+var qweb = core.qweb;
+
+publicWidget.registry.productsRecentlyViewedSnippet = publicWidget.Widget.extend({
+    selector: '.s_wstatistics_products_recently_viewed',
+    xmlDependencies: ['/website_sale/static/src/xml/website_sale_recently_viewed.xml'],
+    events: {
+        'click .js_add_cart': '_onAddToCart',
+    },
+
+    /**
+     * @constructor
+     */
+    init: function () {
+        this._super.apply(this, arguments);
+        this._dp = new concurrency.DropPrevious();
+        this.uniqueId = _.uniqueId('o_carousel_recently_viewed_products_');
+    },
+
+    /**
+     * @override
+     */
+    start: function () {
+        this._dp.add(this._fetch()).then(this._render.bind(this));
+        $(window).resize(() => {
+            if (this.options.products) {
+                this._addCarousel();
+            }
+        });
+        return this._super.apply(this, arguments);
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     */
+    _fetch: function () {
+        return this._rpc({
+            route: '/shop/products/recently_viewed',
+        });
+    },
+
+    /**
+     * @private
+     */
+    _render: function (res) {
+        if (res) {
+            var products = res['products'];
+            this.options = {
+                products: products,
+                uniqueId: this.uniqueId,
+                currency: res['currency'],
+                rating: res['rating'],
+                widget: this,
```

Comment:
this.options.widget.options.widget.options.widget.options... ^^
Why putting the rendering options in `this` instead of passing them to the `_addCarousel` method ? and in fact, why is there a `_addCarousel` method, should not be all this in the `_render` method directly ?